core/vm/instructions.go

```diff
@@ -375,79 +373,42 @@ func opCodeCopy(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([
 		uint64CodeOffset = 0xffffffffffffffff
 	}
 
+	contractAddr := scope.Contract.Address()
 	paddedCodeCopy, copyOffset, nonPaddedCopyLength := getDataAndAdjustedBounds(scope.Contract.Code, uint64CodeOffset, length.Uint64())
 	if interpreter.evm.chainConfig.IsCancun(interpreter.evm.Context.BlockNumber) {
-		scope.Contract.UseGas(touchEachChunksOnReadAndChargeGas(copyOffset, nonPaddedCopyLength, scope.Contract, scope.Contract.Code, interpreter.evm.Accesses, scope.Contract.IsDeployment))
+		scope.Contract.UseGas(touchCodeChunksRangeOnReadAndChargeGas(contractAddr[:], copyOffset, nonPaddedCopyLength, uint64(len(scope.Contract.Code)), interpreter.evm.Accesses))
 	}
 	scope.Memory.Set(memOffset.Uint64(), uint64(len(paddedCodeCopy)), paddedCodeCopy)
 	return nil, nil
 }
 
-// touchChunkOnReadAndChargeGas is a helper function to touch every chunk in a code range and charge witness gas costs
-func touchChunkOnReadAndChargeGas(chunks trie.ChunkedCode, offset uint64, evals [][]byte, code []byte, accesses *state.AccessWitness, deployment bool) uint64 {
-	// note that in the case where the executed code is outside the range of
-	// the contract code but touches the last leaf with contract code in it,
-	// we don't include the last leaf of code in the AccessWitness. The
-	// reason that we do not need the last leaf is the account's code size
-	// is already in the AccessWitness so a stateless verifier can see that
-	// the code from the last leaf is not needed.
-	if code != nil && offset > uint64(len(code)) {
-		return 0
-	}
-	var (
-		chunknr             = offset / 31
-		statelessGasCharged uint64
-	)
-
-	// Build the chunk address from the evaluated address of its whole group
-	var index [32]byte
-	copy(index[:], evals[chunknr/256])
-	index[31] = byte((128 + chunknr) % 256)
-
-	var overflow bool
-	statelessGasCharged, overflow = math.SafeAdd(statelessGasCharged, accesses.TouchAddressOnReadAndComputeGas(index[:]))
-	if overflow {
-		panic("overflow when adding gas")
-	}
-
-	return statelessGasCharged
-}
-
-// touchEachChunksOnReadAndChargeGas is a helper function to touch every chunk in a code range and charge witness gas costs
-func touchEachChunksOnReadAndChargeGas(offset, size uint64, contract *Contract, code []byte, accesses *state.AccessWitness, deployment bool) uint64 {
+// touchCodeChunksRangeOnReadAndChargeGas is a helper function to touch every chunk in a code range and charge witness gas costs
+func touchCodeChunksRangeOnReadAndChargeGas(contractAddr []byte, startPC, size uint64, codeLen uint64, accesses *state.AccessWitness) uint64 {
```

Comment:
This is the new way we do the accounting for the code.
Basically looking at `startPC` and `size` we can know which chunks of the code are being touched.

This means that by doing proper "chunk alignment" logic, we can know which `subTreeIndex` and `subIndex` that corresponds to. 

Might require some careful eyes reviewing it.

Reply:
yup I want to read that twice - but adding a couple of tests would do the trick as well